mage_ai/api/resources/PipelineResource.py

```diff
@@ -414,6 +418,40 @@ def retry_pipeline_runs(pipeline_runs):
             for run in pipeline_runs:
                 retry_pipeline_run(run)
 
+        @safe_db_query
+        def query_incomplete_block_runs(pipeline_uuid: str):
+            a = aliased(PipelineRun, name='a')
+            b = aliased(BlockRun, name='b')
+            columns = [
+                b.id,
+                b.status,
+                b.pipeline_run_id,
+                a.status,
+                a.pipeline_uuid,
+            ]
+            result = (
+                PipelineRun.
+                select(*columns).
+                join(b, a.id == b.pipeline_run_id).
+                filter(a.pipeline_uuid == pipeline_uuid).
+                filter(b.status != BlockRun.BlockRunStatus.COMPLETED)
+            ).all()
+
+            return result
+
+        def retry_incomplete_block_runs(pipeline_uuid: str):
+            incomplete_block_run_results = query_incomplete_block_runs(pipeline_uuid)
+            block_run_ids = [r[0] for r in incomplete_block_run_results]
+            pipeline_run_ids = list(set([r[2] for r in incomplete_block_run_results]))
+            BlockRun.batch_update_status(
+                block_run_ids,
+                BlockRun.BlockRunStatus.INITIAL,
+            )
+            PipelineRun.batch_update_status(
+                pipeline_run_ids,
+                PipelineRun.PipelineRunStatus.RUNNING,
```

Comment:
We can actually update the status to INITIAL so that we can enforce concurrency control on pipeline runs.